cdc/sink/flowcontrol/flow_control.go

```diff
@@ -134,48 +160,55 @@ func (c *TableFlowController) enqueueSingleMsg(msg *model.PolymorphicEvent, size
 	// 1. Processing a new txn with different commitTs.
 	if e = c.queueMu.queue.Back(); e == nil || lastCommitTs < commitTs {
 		atomic.StoreUint64(&c.lastCommitTs, commitTs)
-		c.queueMu.queue.PushBack(&txnSizeEntry{
-			startTs:  msg.StartTs,
-			commitTs: commitTs,
-			size:     size,
-			rowCount: 1,
-		})
-		c.batchGroupCount = 1
-		msg.Row.SplitTxn = true
+		c.resetBatch(lastCommitTs, commitTs)
+		c.addEntry(msg, size)
 		return
 	}
 
 	// Processing txns with the same commitTs.
 	txnEntry := e.(*txnSizeEntry)
 	if txnEntry.commitTs != lastCommitTs {
-		log.Panic("got wrong commitTs from deque, report a bug",
+		log.Panic("got wrong commitTs from deque in flow control, report a bug",
 			zap.Uint64("lastCommitTs", c.lastCommitTs),
 			zap.Uint64("commitTsInDeque", txnEntry.commitTs))
 	}
 
 	// 2. Append row to current txn entry.
-	if txnEntry.startTs == msg.Row.StartTs &&
-		txnEntry.rowCount < maxRowsPerTxn && txnEntry.size < maxSizePerTxn {
+	if txnEntry.batchID == c.batchID && txnEntry.startTs == msg.Row.StartTs &&
+		txnEntry.rowCount < c.maxRowsPerTxn && txnEntry.size < c.maxSizePerTxn {
 		txnEntry.size += size
 		txnEntry.rowCount++
 		return
 	}
 
 	// 3. Split the txn or handle a new txn with the same commitTs.
+	if c.batchGroupCount >= c.batchSize {
+		_ = callback()
```

Comment:
Should reset batch here ?

Reply:
Yes, batch is reset inside [callback](https://github.com/pingcap/tiflow/pull/5712/files#diff-cfe8ccb173f8b93acc1b97cc5825fdb565d539634714fadffacdf18ea4bc8b26R100).